backend/de.metas.business.rest-api-impl/src/main/java/de/metas/rest_api/v2/project/budget/BudgetProjectRestService.java

```diff
@@ -0,0 +1,301 @@
+import de.metas.project.budget.BudgetProject;
+import de.metas.project.budget.BudgetProjectData;
+import de.metas.project.budget.BudgetProjectRepository;
+import de.metas.project.budget.BudgetProjectResource;
+import de.metas.project.budget.BudgetProjectResourceData;
+import de.metas.project.budget.UpsertBudgetProjectRequest;
+import de.metas.project.workorder.data.ProjectQuery;
+import de.metas.resource.ResourceGroupId;
+import de.metas.rest_api.utils.IdentifierString;
+import de.metas.rest_api.utils.MetasfreshId;
+import de.metas.uom.UomId;
+import de.metas.user.UserId;
+import de.metas.util.Check;
+import de.metas.util.Services;
+import de.metas.util.lang.ExternalId;
+import de.metas.util.web.exception.InvalidIdentifierException;
+import de.metas.util.web.exception.MissingPropertyException;
+import de.metas.util.web.exception.MissingResourceException;
+import lombok.NonNull;
+import org.adempiere.ad.trx.api.ITrxManager;
+import org.adempiere.exceptions.AdempiereException;
+import org.springframework.stereotype.Service;
+
+import javax.annotation.Nullable;
+import java.time.LocalDateTime;
+import java.time.ZoneId;
+import java.util.Optional;
+
+@Service
+public class BudgetProjectRestService
+{
+	private final ITrxManager trxManager = Services.get(ITrxManager.class);
+	private final IOrgDAO orgDAO = Services.get(IOrgDAO.class);
+	private final ICurrencyBL currencyBL = Services.get(ICurrencyBL.class);
+
+	private final BudgetProjectRepository budgetProjectRepository;
+	private final ProjectTypeRepository projectTypeRepository;
+	private final BudgetProjectJsonConverter budgetProjectJsonConverter;
+
+	public BudgetProjectRestService(
+			@NonNull final BudgetProjectRepository budgetProjectRepository,
+			@NonNull final ProjectTypeRepository projectTypeRepository,
+			@NonNull final BudgetProjectJsonConverter budgetProjectJsonConverter)
+	{
+		this.budgetProjectRepository = budgetProjectRepository;
+		this.projectTypeRepository = projectTypeRepository;
+		this.budgetProjectJsonConverter = budgetProjectJsonConverter;
+	}
+
+	@NonNull
+	public JsonBudgetProjectResponse getBudgetProjectDataById(@NonNull final ProjectId projectId)
+	{
+		final BudgetProject projectData = budgetProjectRepository.getOptionalById(projectId)
+				.orElseThrow(() -> MissingResourceException.builder()
+						.resourceName("Budget Project")
+						.resourceIdentifier(String.valueOf(projectId.getRepoId()))
+						.build());
+
+		return toJsonBudgetProjectResponse(projectData);
+	}
+
+	@NonNull
+	public JsonBudgetProjectUpsertResponse upsertBudgetProject(@NonNull final JsonBudgetProjectUpsertRequest request)
+	{
+		return trxManager.callInNewTrx(() -> upsertBudgetProjectWithinTrx(request));
+	}
+
+	@NonNull
+	private JsonBudgetProjectUpsertResponse upsertBudgetProjectWithinTrx(@NonNull final JsonBudgetProjectUpsertRequest request)
+	{
+		final SyncAdvise budgetProjectSyncAdvise = request.getSyncAdvise();
+		if (budgetProjectSyncAdvise == null)
```

Comment:
why not making syncAdvice mandatory and fail when JsonBudgetProjectUpsertRequest is constructed?

Reply:
bc JsonBudgetProjectUpsertRequest is not Immutable && we build it via setters in camel so I think it will be weird to give some fields via constructor/builder and others via setters.. (we use setters in order to know which fields were actually set - i.e. `isNameSet` )

Reply:
understand